### PR TITLE
[WIP] [STILL REVIEW] HTTP/1.1 and gRPC Multiplexer Inbound

### PR DESCRIPTION
**Pull request by peter-edge**
This is a POC for a HTTP/1.1 and gRPC multiplexer inbound using https://github.com/soheilhy/cmux. The package `transport/x/grpchttp` provides a `transport.Inbound` that starts both a `*http.Inbound` and `*grpc.Inbound`.

This is not ready to merge, I need to carefully go over cmux, and make sure I am doing this correctly. Additionally, there's a deadlock on `Stop()` when stopping the `*http.Inbound`.

Other than that issue, this has been tested to work end-to-end for both http and grpc outbounds. Additionally, no performance penalty is observed:

```
# first run
$ go test -run Integration -bench Integration -benchmem -parallel 1 ./encoding/protobuf/testing
goos: darwin
goarch: amd64
pkg: go.uber.org/yarpc/encoding/protobuf/testing
BenchmarkIntegrationYARPC/http/Get-8     	   10000	    113716 ns/op	   11314 B/op	     171 allocs/op
BenchmarkIntegrationYARPC/tchannel/Get-8 	   10000	    105353 ns/op	   12742 B/op	     146 allocs/op
BenchmarkIntegrationYARPC/grpc/Get-8     	   10000	    139945 ns/op	   11738 B/op	     210 allocs/op
BenchmarkIntegrationYARPC/grpchttp/Get-8 	   10000	    140608 ns/op	   11735 B/op	     210 allocs/op
BenchmarkIntegrationGRPC/grpc/Get-8      	   10000	    128226 ns/op	    9331 B/op	     170 allocs/op
BenchmarkIntegrationGRPC/grpchttp/Get-8  	   10000	    144027 ns/op	    9335 B/op	     170 allocs/op

# second run
$ go test -run Integration -bench Integration -benchmem -parallel 1 ./encoding/protobuf/testing
goos: darwin
goarch: amd64
pkg: go.uber.org/yarpc/encoding/protobuf/testing
BenchmarkIntegrationYARPC/http/Get-8     	   10000	    111126 ns/op	   11318 B/op	     171 allocs/op
BenchmarkIntegrationYARPC/tchannel/Get-8 	   10000	    105948 ns/op	   12706 B/op	     146 allocs/op
BenchmarkIntegrationYARPC/grpc/Get-8     	   10000	    145186 ns/op	   11741 B/op	     210 allocs/op
BenchmarkIntegrationYARPC/grpchttp/Get-8 	   10000	    140417 ns/op	   11735 B/op	     210 allocs/op
BenchmarkIntegrationGRPC/grpc/Get-8      	   10000	    127535 ns/op	    9330 B/op	     170 allocs/op
BenchmarkIntegrationGRPC/grpchttp/Get-8  	   10000	    129429 ns/op	    9330 B/op	     170 allocs/op
```

The `YARPC/grpchttp` benchmark above is using a `*grpc.Outbound` talking to a `*grpchttp.Inbound`, and the `GRPC/grpchttp` benchmark is using a grpc-go `*grpc.ClientConn` talking to a `*grpchttp.Inbound`.

This relies on https://github.com/soheilhy/cmux/pull/50.

### CODE REVIEW

**Comment by peter-edge** [comment]
I'd like to revisit this and see if we can expose this in an experimental package. I have a request for this and it'd be cool to play with it.

File: transport/x/grpchttp/grpchttp.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+	grpcTransport := yarpcgrpc.NewTransport(i.options.grpcTransportOptions...)
+	grpcInbound := grpcTransport.NewInbound(grpcListener, i.options.grpcInboundOptions...)
+	grpcInbound.SetRouter(i.router)
+
+	httpTransport := yarpchttp.NewTransport(i.options.httpTransportOptions...)
+	httpInbound := httpTransport.NewInboundForListener(httpListener, i.options.httpInboundOptions...)
+	httpInbound.SetRouter(i.router)
+
+	if err := grpcTransport.Start(); err != nil {
+		_ = i.listener.Close()
```
[the remark this answers is not in the document]
Comment (abhinav) [reply transport/x/grpchttp/grpchttp.go]: Also, who started this listener? Was it us or the user? If the user started it, we probably want to document (when polishing this) that we take ownership of the listener and will close it when it's time.

File: transport/x/grpchttp/grpchttp.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		i.listener = nil
+		return err
+	}
+	if err := grpcInbound.Start(); err != nil {
+		_ = grpcTransport.Stop()
+		_ = i.listener.Close()
```
Comment (abhinav) [inline transport/x/grpchttp/grpchttp.go]: once Transport is added, that should be responsible for stopping the grpc and http transports.

File: transport/http/inbound.go [hunk 3]
```diff
@@ -83,17 +83,34 @@ func (t *Transport) NewInbound(addr string, opts ...InboundOption) *Inbound {
 	return i
 }
 
+// NewInboundForListener builds a new HTTP inbound that listens on the given
+// listener and sharing this transport.
+func (t *Transport) NewInboundForListener(listener net.Listener, opts ...InboundOption) *Inbound {
```
[the remark this answers is not in the document]
Comment (peter-edge) [reply transport/http/inbound.go]: I'd disagree on this one - `addr string` means nothing in this context, so it's confusing that I could pass an address and it does nothing.

[the remark this answers is not in the document]
Comment (abhinav) [reply transport/http/inbound.go]: Oh, I missed that we have a required `addr`. Makes sense.

**Pull request closed without merging**